### PR TITLE
SDMMC SD Card Drivers

### PR DESCRIPTION
**Pull request by lekangwang**
## Description

*What was completed, changed, or updated?* 

- Added FATFS third party libraries
- Enabled and configure SDMMC to 4-bit data mode
- Update CMakeList.txt to compile FATFS files
- Added low level drivers that calls HAL functions to interact with SD card via SDMMC board peripheral
- Created 2 thread-safe high level functions to read/write to SD card, (file names are limited to 8.3 format)

<br/>

*Why was this done (if applicable)?*

- For logging purposes (i.e. crash report)

<br/>

## Testing

*What **manual** tests were used to validate the code?*

Tested on 16 and 32 GB card
- Can write the same message 100 times to a file
Screenshot 2024-04-11 at 10 01 12 PM

- Reading the first line, data matches as expected
Screenshot 2024-04-11 at 10 00 06 PM

- Multithreaded write to same file is handled correctly
Screenshot 2024-05-30 at 11 53 18 PM

<br/>

*What **unit** tests were used to validate the code?*

- Low level drivers passes unit tests written by the FATFS author (http://elm-chan.org/fsw/ff/res/app4.c)

<br/>

## Documentation

*Milestone number and name:* M2

*Link to Asana task:* https://app.asana.com/0/1203458353737758/1204645027312218/f

*Link to Confluence documentation:* https://uwarg-docs.atlassian.net/wiki/spaces/ZP/pages/2251620424/SDMMC+SD+Card

<br/>

## Reminders

- [x] Add reviewers to the PR

- [x] Mention the PR in the appropriate discord channel

### CODE REVIEW

File: Boardfiles/nucleol552zeq/Core/Src/i2c.c [hunk 1]
```diff
@@ -82,7 +82,7 @@ void MX_I2C2_Init(void)
 
   /* USER CODE END I2C2_Init 1 */
   hi2c2.Instance = I2C2;
-  hi2c2.Init.Timing = 0x40505681;
+  hi2c2.Init.Timing = 0x20505885;
```
Comment (antholuo) [inline Boardfiles/nucleol552zeq/Core/Src/i2c.c:85]: Why did we need to change this timing? Or was it just a i2c configuration changed and this was part of the auto-generated files?

File: Boardfiles/nucleol552zeq/Core/Src/sdmmc.c [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    PC10     ------> SDMMC1_D2
+    PC11     ------> SDMMC1_D3
+    PC12     ------> SDMMC1_CK
+    PD2     ------> SDMMC1_CMD
+    */
+    GPIO_InitStruct.Pin = GPIO_PIN_8|GPIO_PIN_9|GPIO_PIN_10|GPIO_PIN_11
+                          |GPIO_PIN_12;
+    GPIO_InitStruct.Mode = GPIO_MODE_AF_PP;
+    GPIO_InitStruct.Pull = GPIO_NOPULL;
+    GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
```
Comment (antholuo) [inline Boardfiles/nucleol552zeq/Core/Src/sdmmc.c:90]: Does our speed need to be very high? This tends to cause ringing and *can* cause bad data.

Would advise to bring it back down to FREQ_MEDIUM or FREQ_LOW.

File: Boardfiles/nucleol552zeq/Core/Src/sdmmc.c [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+    GPIO_InitStruct.Pin = GPIO_PIN_2;
+    GPIO_InitStruct.Mode = GPIO_MODE_AF_PP;
+    GPIO_InitStruct.Pull = GPIO_NOPULL;
+    GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
+    GPIO_InitStruct.Alternate = GPIO_AF12_SDMMC1;
+    HAL_GPIO_Init(GPIOD, &GPIO_InitStruct);
+
+    /* SDMMC1 interrupt Init */
+    HAL_NVIC_SetPriority(SDMMC1_IRQn, 5, 0);
```
Comment (antholuo) [inline Boardfiles/nucleol552zeq/Core/Src/sdmmc.c:102]: This isn't something that we need to consider yet, but we should probably start thinking about our task / interrupt priorities....

**Review by antholuo — APPROVED** [review]
Changes mostly look good, especially if tested working.

I'm curious to see how it plays with freertos task scheduling and if that might cause problems with data stalling or weird interrupt issues. Might be worth trying to schedule two threads and having one thread write to the SD card while the other one does busy work?

File: Boardfiles/nucleol552zeq/Core/Src/usart.c [hunk 4]
```diff
@@ -46,7 +44,7 @@ void MX_LPUART1_UART_Init(void)
 
   /* USER CODE END LPUART1_Init 1 */
   hlpuart1.Instance = LPUART1;
-  hlpuart1.Init.BaudRate = 209700;
+  hlpuart1.Init.BaudRate = 230400;
```
Comment (HardyYu) [inline Boardfiles/nucleol552zeq/Core/Src/usart.c:47]: This is a similar issue as the i2c, the baudrate is changed wiredly

Comment (lekangwang) [reply Boardfiles/nucleol552zeq/Core/Src/usart.c:47]: I did this so I could print stuff on my terminal on my laptop, will change this back later

File: Boardfiles/nucleol552zeq/Core/Src/sdmmc.c (hunk 2 above)
Comment (lekangwang) [reply Boardfiles/nucleol552zeq/Core/Src/sdmmc.c:90]: Changed to FREQ_MEDIUM

File: Boardfiles/nucleol552zeq/Core/Src/i2c.c (hunk 1 above)
Comment (lekangwang) [reply Boardfiles/nucleol552zeq/Core/Src/i2c.c:85]: I was using PLL clock for the SDMMC peripheral, and I changed the PLL freq to be slow enough for the SD card. I changed this so I'm not using PLL anymore and instead Clock48 with a clock divider factor of 8 for SDMMC peripheral.

**Review by HardyYu — APPROVED** [review]
Wow! Having DMA and thread guard implemented is really cool. Very impressive! This pr is kinda large and couldn't finish reviewing it today. I will leave my approvement here since this  code was tested to be working. Good work!

File: Drivers/sd_card/src/bsp_driver_sd.c [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+/**
+  * @brief RTOS-aware delay
+  * @param delay: number of ticks to delay
+  * @retval None
+  */
+void HAL_Delay(uint32_t delay)
+{
+  osDelay(delay);
```
[the remark this answers is not in the document]
Comment (lekangwang) [reply Drivers/sd_card/src/bsp_driver_sd.c:172]: its used in ZeroPilot-SW-3.5/Boardfiles/nucleol552zeq/Drivers/STM32L5xx_HAL_Driver/Src/stm32l5xx_hal_sd.c for function `HAL_SD_InitCard` which is called by `HAL_SD_Init` which is called by `BSP_SD_Init`

**Pull request merged**